From 9172d7e36946288862fdff4a9ca0b3506acd28c7 Mon Sep 17 00:00:00 2001
From: Haohui Mai <wheat9@apache.org>
Date: Wed, 4 Jan 2017 21:01:23 -0800
Subject: [PATCH 2287/2748] HDFS-11280. Allow WebHDFS to reuse HTTP
 connections to NN. Contributed by Zheng Shao.

(cherry picked from commit 977fb0c52468d3065c7e887580c050ac04b403e3)

Change-Id: Icf97d35d7eac8a1994ac031026ef165591959d97
---
 .../apache/hadoop/hdfs/web/WebHdfsFileSystem.java  |    9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/web/WebHdfsFileSystem.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/web/WebHdfsFileSystem.java
index 08280fa..f33e3cf 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/web/WebHdfsFileSystem.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/web/WebHdfsFileSystem.java
@@ -592,6 +592,8 @@ protected HttpURLConnection connect(URL url) throws IOException {
           url = new URL(conn.getHeaderField("Location"));
           redirectHost = url.getHost() + ":" + url.getPort();
         } finally {
+          // TODO: consider not calling conn.disconnect() to allow connection reuse
+          // See http://tinyurl.com/java7-http-keepalive
           conn.disconnect();
         }
       }
@@ -819,7 +821,9 @@ final T getResponse(HttpURLConnection conn) throws IOException {
         }
         throw ioe;
       } finally {
-        conn.disconnect();
+        // Don't call conn.disconnect() to allow connection reuse
+        // See http://tinyurl.com/java7-http-keepalive
+        conn.getInputStream().close();
       }
     }
     
@@ -865,6 +869,9 @@ public void close() throws IOException {
             try {
               validateResponse(op, conn, true);
             } finally {
+              // This is a connection to DataNode.  Let's disconnect since
+              // there is little chance that the connection will be reused
+              // any time soonl
               conn.disconnect();
             }
           }
-- 
1.7.9.5

